frontend/dockerfile/shell/lex.go

```diff
@@ -394,6 +396,49 @@ func (sw *shellWord) processDollar() (string, error) {
 				return "", errors.Errorf("%s: %s", name, message)
 			}
 			return value, nil
+		case '%', '#':
+			// %/# matches the shortest pattern expansion, %%/## the longest
+			match_longest := false
+			if word[0] == byte(ch) {
+				match_longest = true
+				word = word[1:]
+			}
+
+			// regular expressions can't handle finding the shortest rightmost
+			// string so we reverse both search space and pattern to convert it
+			// to a leftmost search in both cases
+			if ch == '%' {
+				value = bidi.ReverseString(value)
+				word = bidi.ReverseString(word)
+			}
+
+			// generate a regular expression for the provided pattern by escaping all
+			// meta characters first
+			pattern := "^" + regexp.QuoteMeta(word)
+			// \? (after escaping) should match a single character: .
+			pattern = strings.Replace(pattern, "\\?", ".", -1)
+			// \* (after escaping) should match the shortest (.*?) or longest (.*) substring
+			if match_longest {
+				pattern = strings.Replace(pattern, "\\*", ".*", -1)
+			} else {
+				pattern = strings.Replace(pattern, "\\*", ".*?", -1)
+			}
+
+			re, err := regexp.Compile(pattern)
```

Comment:
I don't think this is the best way to implement this algorithm. From my experience, regular expressions that are created by using `QuoteMeta` and `strings.Replace` generally become very fragile and hard to reason about.

I understand wanting to use the `regexp` package for this though.

I think something I might recommend is to implement a scanner for the [pattern matching notation](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13) and generate a regular expression from that.

Just mocking this out, but probably something like this: https://go.dev/play/p/Su08PINS_IY

This is usually a lot more reliable of a way to translate a string to a regular expression.

Reply:
I think regular expressions are the best way to handle greedy and non-greedy matches with any number of wildcards while keeping the binary size reasonable. One problem I encountered was that the pattern is already parsed for escapes, and after enabling the RawEscapes option the `\}` needs to be handled differently.

The new implementation is close to your PoC, but uses a scanner for the `Peek()`-functionality.